File: extras/cola/cola.rb
require_relative '../code'

TiempoMinimoAtencion    = 2
TiempoMinimoObservacion = 5

def promedio(a, b)
    return 0 if b.cantidad == a.cantidad
    ((b.hora - a.hora) / (b.cantidad - a.cantidad).to_f).abs.round(2)
end

class Cola 
    attr_accessor :nombre, :reloj, :registros
    
    def initialize(nombre='Cola')
        self.nombre = nombre
        self.reloj  = nil 
        self.registros = []
    end

    def entrar
        registrar(+1) 
    end

    def salir
        registrar(-1)
    end

    def configurando?
        primero = entradas.first 
        !primero || (self.hora - primero.hora >= TiempoMinimoObservacion)
    end

    def editando?
        ultimo = entradas.last 
        !ultimo || (self.hora - ultimo.hora >= TiempoMinimoAtencion)
    end

    def hora 
        self.reloj || Time.new
    end

    def entradas
        permanentes = []
        reducir(self.registros) do | anterior, actual |
            if actual 
                permanentes << anterior if actual.hora - anterior.hora >= TiempoMinimoAtencion
            else 
                permanentes << anterior if self.hora - anterior.hora >= TiempoMinimoAtencion
            end
        end

        salida = []
        reducir(permanentes) do | anterior, actual |
            next unless actual
            if salida.count == 0
                salida << anterior if actual.hora - anterior.hora >= TiempoMinimoObservacion
            end 
            salida << actual   if actual.cantidad < anterior.cantidad 
        end
        salida 
    end

    def muestra(cantidad=1000)
        lista = self.entradas.clone
        cantidad.times.map do
            i = rand(lista.count)
            j = rand(lista.count)
            redo if i == j
            promedio(lista[i], lista[j])
        end
    end

    def registrar(cantidad)
        cantidad += registros.last ? registros.last.cantidad : 0
        self.registros << { cantidad: cantidad, hora: self.hora } if cantidad >= 0
    end

    def avanzar(tiempo=1)
        self.reloj ||= 0
        self.reloj += tiempo if tiempo > 0
    end

    def mostrar(lista, texto='')
        puts " #{nombre} (#{self.hora}s, #{lista.count}) [#{texto}] #{configurando? ? :configurando : ''} #{editando? ? :editando : ''} ".titulo do 
            lista.each{|e| puts ' • %2is > %2i ' % [e.hora, e.cantidad] }
        end
        puts ''
    end
    
    def reducir(lista)
        pares = lista.zip(lista[1..-1])
        pares.each{|anterior, actual| yield( anterior, actual ) }
    end
end

# ---

def Cola(nombre, cantidad=0, &bloque)
    tmp = Cola.new( nombre )
    tmp.avanzar(0)
    cantidad.times{ tmp.entrar }
    tmp.instance_eval( &bloque )
    tmp 
end

if __FILE__ == $0
    puts ' Simulacion '.pad(100).error
    Cola 'Polo Norte' do 
        entrar
        entrar
        entrar
        entrar
        entrar
        avanzar 10 #5 
        salir
        salir       
        avanzar 30  #3
        salir 
        avanzar 10  #2
        
        pp muestra(100).promedio.round(1)
    end
 end

File: extras/cola/cola.js
const TiempoMinimoAtencion    = 2;
const TiempoMinimoObservacion = 5;

Array.prototype.last = function() {
    return this[this.length - 1];
}

Array.prototype.first = function() {
    return this[0];
}

Array.prototype.promedio = function(){
    return this.reduce((valor, suma) => suma + valor) / this.length;
}

function promedio(a, b){
    if( b.cantidad == a.cantidad) { 
        return 0;
    }
    return Math.abs((b.hora - a.hora) / (b.cantidad - a.cantidad));
}

function rand(maximo){
    return Math.floor(Math.random() * maximo); 
}


class Cola {

    constructor (nombre){
        this.nombre = nombre;
        this.reloj  = 0 ;
        this.registros = [];
    }

    entrar(){
        return this.registrar(+1); 
    }

    salir(){
        return this.registrar(-1);
    }

    configurando() {
        primero = this.entradas().first();
        return !primero || (this.hora() - primero.hora >= TiempoMinimoObservacion)
    }

    editando(){
        ultimo = this.entradas().last();
        return !ultimo || (this.hora() - ultimo.hora >= TiempoMinimoAtencion)
    }

    hora() {
        return this.reloj;
    }

    entradas() {
        let permanentes = [];
        this.reducir(this.registros, (anterior, actual) => {
            if( actual != undefined ){
                if( actual.hora - anterior.hora >= TiempoMinimoAtencion) {
                    permanentes.push(anterior);
                }
            } else {
                if (this.hora() - anterior.hora >= TiempoMinimoAtencion){
                    permanentes.push(anterior);
                }
            }
        });

        let salida = [];
        this.reducir(permanentes, (anterior, actual) => {
            if(actual != undefined){ 
                if(salida.length == 0){
                    if ( actual.hora - anterior.hora >= TiempoMinimoObservacion ){
                        salida.push(anterior);
                    }
                } 
                if( actual.cantidad < anterior.cantidad ) {
                        salida.push(actual);
                }
            }
        });
        return salida;
    }

    muestra(cantidad=100){
        let lista = this.entradas();
        let promedios = [];
        let a, b;
        for(var i = 0; i < cantidad; i++){
            do {
                a = rand(lista.length);
                b = rand(lista.length);
            } while(a == b);
            promedios.push( promedio(lista[a], lista[b]) );
        }
        return promedios;
    }

    registrar(cantidad){
        cantidad += this.registros.last() ? this.registros.last().cantidad : 0;
        if(cantidad >= 0){
            this.registros.push({ cantidad: cantidad, hora: this.hora() });
        }
    }

    avanzar(tiempo=1){
        this.reloj = this.reloj || 0;
        if(tiempo > 0){
            this.reloj += tiempo;
        }
    }

    mostrar(lista, texto=''){
        console.log(` ${nombre} (${this.hora()}s, ${lista.length}) [${texto}] ${this.configurando ? 'configurando' : ''} ${this.editando ? 'editando' : ''} `)
        lista.forEach( e => console.log(`   • ${e.hora} > ${e.cantidad}`) );
        console.log("");
    }
  
    reducir(lista, acciones){
        for(var i = 0; i < lista.length; i++){
            var anterior = lista[i];
            var actual   = lista[i+1];
            acciones( anterior, actual);
        }
    }

}

var cola = new Cola("Polo Norte");
cola.entrar();
cola.entrar();
cola.entrar();
cola.entrar();
cola.entrar();
cola.avanzar(10);   // 5 
cola.salir();
cola.salir();       // 3
cola.avanzar(30);
cola.salir();       // 2
cola.avanzar(10);
console.log(cola.muestra(100).promedio());

console.log("Todo Ok")


// def Cola(nombre, cantidad=0, &bloque)
//     tmp = Cola.new( nombre )
//     tmp.avanzar(0)
//     cantidad.times{ tmp.entrar }
//     tmp.instance_eval( &bloque )
//     tmp 
// end
